Refactor BaseWireframe (change `viewController` property to `getDeallocatableViewController` method)

terzicluka opened the pull request:
This pr has changed the `BaseWireframe` implementation. Since the old implementation, which had the `viewController` property, introduced issues on multiple projects which were hard to debug, we decided to implement a solution that would be easier to debug and less error-prone. 

The `viewController` property is replaced with the `getDeallocatableViewController` method which suggests that the `viewController` instance will be deallocated if the caller doesn't hold a strong reference to it at the call site. Also, we introduced a fatalError to the `getDeallocatableViewController` method in order to provide a better error description if the user of the method doesn't use the method as it was intended.

Truba commented:
> Since the old implementation, which had the viewController property, introduced issues on multiple projects which were hard to debug

Could you please share one of the issues?

terzicluka commented:
@Truba Maybe I have put it in the wrong words but the `BaseWireframe` didn't produce issues but the developer who used the `viewController` property from `BaseWireframe`.

The concrete issue in one of our projects:
```swift
        if wireframe.viewController is UINavigationController {
            navigationController?.present(viewController: wireframe.viewController)
            return
        }
        let navController = SomeNavigationController()
        navController.setRootWireframe(wireframe) // setViewControllers([wireframe.viewController], animated: animated)
        navigationController?.present(viewController: navController)
```

As you can see, in this implementation you would access the `viewController` property twice without storing a strong reference at the call site which means that the ViewController deinitializes since its reference counter gets decremented to 0 after the first access to `wireframe.viewController`.

The issue didn't manifest on the project for years in production. When using `UIKit's` regular `UIViewController` this chunk of code worked, but since we introduced `SwiftUI` and  `LazyHostingViewController` the issue manifested.

In the process of debugging the crash, we had a hard time implementing a fix since we didn't even consider that we should keep a strong reference to the `viewController` at the call site. Also, Xcode didn't give us any useful error messages after the app crashed which also didn't navigate us to the real root cause of the problem. After a lot of debugging we managed to implement a fix for this bug:
```swift
        /// `vc` is needed to hold a strong reference to wireframes `viewController` property because accessing `wireframe.viewController` sets the `_temporaryStoredViewController` property to `nil`. Since we need to access the `viewController` twice, the app will crash if we don't hold a strong reference inside of the scope of this method.
        let vc = wireframe.viewController

        if vc is UINavigationController {
            navigationController?.present(viewController: vc)
            return
        }
        let navController = SomeNavigationController()
        navController.setRootWireframe(wireframe)
        navigationController?.present(viewController: navController)
```
Also, I was informed that similar issues happened in the past. We decided to make it harder for the developer to produce this bug while using the `BaseWireframe` implementation. We rewrote the `viewController` property to a method that has a more descriptive name about what could happen when you use it. Also, we added a `fatalError` inside the method so if the app crashes you could get a better error message about what happened and you could easily fix the issue since you will instantly see what was the root cause of the issue.

Truba commented:
@terzicluka I see now. 

I can get behind adding fatal error instead of unowned for debugging purposes.

But doing a rename from `viewController` to `getDeallocatableViewController` will have to be a major version update and you guys would have to update all the templates. Then everyone would have to update their code as well. IMHO the rename isn't worth it.

p.s. If I can propose a nicer refactor of your code.
```swift 
        if let nextNavigationController = wireframe.viewController as? UINavigationController {
            navigationController?.present(viewController: nextNavigationController)
            return
        }
        let navController = SomeNavigationController()
        navController.setRootWireframe(wireframe) // setViewControllers([wireframe.viewController], animated: animated)
        navigationController?.present(viewController: navController)
```
in cases as such I would always prefer `if/guard let` than doing `is`. It shows intent more explicitly.

Truba reviewed (changes requested):
As mentioned you either need to remove the rename or update all of the templates and increase the major version.

terzicluka commented:
I have reverted the property rename but decided to keep the fatal error for debugging purposes.

Truba commented:
👍  
Please publish as a 4.1 with the addition to readme.
